receiver/m365receiver/client.go

```diff
@@ -0,0 +1,424 @@
+	if resp.StatusCode != 200 {
+		//get error code
+		respErr, err := m.readInErr(resp)
+		if err != nil {
+			return []string{}, err
+		}
+		if respErr.Error.Code == "InvalidAuthenticationToken" {
+			return []string{}, fmt.Errorf("access token invalid")
+		}
+		return []string{}, fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	csvReader := csv.NewReader(resp.Body)
+
+	//parse out 2nd line & return csv data
+	_, err = csvReader.Read()
+	if err != nil {
+		return []string{}, err
+	}
+	data, err := csvReader.Read()
+	if err != nil {
+		if err == io.EOF { //no data in report, scraper should still run
+			return []string{}, nil
+		}
+		return []string{}, err
+	}
+
+	return data, nil
+}
+
+// function for getting log data
+func (m *m365Client) GetJSON(ctx context.Context, endpoint string, end string, start string) (logData, error) {
+	// make request to Audit endpoint
+	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
+	if err != nil {
+		return logData{}, err
+	}
+	req.Header.Set("Authorization", "Bearer "+m.token)
+	q := req.URL.Query()
+	q.Add("startTime", start)
+	q.Add("endTime", end)
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return logData{}, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot error code
+	if resp.StatusCode != 200 {
+		respErr, err := m.readInErr(resp)
+		if err != nil {
+			return logData{}, err
+		}
+		if respErr.Message == "Authorization has been denied for this request." {
+			return logData{}, fmt.Errorf("authorization denied")
+		}
+		return logData{}, fmt.Errorf("got non 200 status code from request, got %d", resp.StatusCode)
+	}
+
+	// check if contentUri field if present
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return logData{}, err
+	}
+	if string(body) == "[]" { // if body is empty, no new log data available
+		return logData{}, nil
+	}
+
+	// extract contentUri field
+	var lr []logResp
+	err = json.Unmarshal(body, &lr)
+	if err != nil {
+		return logData{}, err
+	}
+
+	// read in json
+	body, err = m.followLink(ctx, &lr[0])
+	if err != nil {
+		return logData{}, err
+	}
+	var data logData
+	err = json.Unmarshal(body, &data.logs)
+	if err != nil {
+		return logData{}, err
+	}
+	data.body = m.parseBody(body)
+
+	return data, nil
+}
+
+func (m *m365Client) StartSubscription(endpoint string) error {
+	req, err := http.NewRequest("POST", endpoint, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", "Bearer "+m.token)
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// troubleshoot error handling, mainly sub already enabled
+	// no error if sub already enabled, not troubleshooting stale token
+	// only called while code is synchronous right after a GetToken call
+	// if token is stale, regenerating it won't fix anything
+	if resp.StatusCode != 200 {
```

Comment:
Could refactor error handling out of the three functions to clean up client a bit.